### PR TITLE
[FEATURE] Add DINO Score metric with tests and docs 

### LINKED ISSUE #318: Add DINO Score
## ✨ What You’ll Do
Adding more metrics makes Pruna’s evaluations more insightful, so we’re excited to welcome **DINO Score** ([paper](https://arxiv.org/abs/2208.12242), [repo](https://github.com/google/dreambooth)) into our metric lineup! 🎉

DINO Score uses embeddings from a pretrained **DINO ViT-S/16** backbone and computes the mean cosine similarity between generated images and their reference images. This helps measure **subject fidelity** and **visual consistency**, complementing existing metrics like FID, KID, and CLIPScore.

---

## 🛠️ Dependency
Add `timm` to `pyproject.toml` so the DINO ViT-S/16 model can be imported.

---

## 📐 Metric Class
For the full deep-dive on Pruna metrics (class structure, state handling, tests, registration), check out our guide: [Customize a Metric](https://docs.pruna.ai/en/stable/docs_pruna/user_manual/customize_metric.html)

**Quick Overview**
1. **File & Class**
   * Create `pruna/evaluation/metrics/metric_dino_score.py`
   * Inherit from `StatefulMetric`
2. **Configuration**
   * Choose a sensible `call_type` and set `metric_name = "dino_score"`
   * Define and register any internal state via `self.add_state(...)`
3. **Implement Core Methods**
   * `update(x, gt, outputs)`
     * Receive inputs, ground truths, and model output
     * Extract embeddings from DINO ViT-S/16 ([CLS] token)
     * Compute cosine similarities between generated and reference embeddings
     * Accumulate similarities in your state variables
   * `compute()`
     * Aggregate (e.g., average all scores)
     * Return result with `name="dino_score"` and the aggregated float
4. **Registration**
   * Use the `@MetricRegistry.register("dino_score")` decorator so users can invoke the metric by name
5. **Testing & Docs**
   * Write unit tests under `tests/evaluation/metrics/test_dino_score.py`
   * Update `user_manual/evaluation.rst` to mention `"dino_score"`

---

## ✅ Acceptance Criteria
* **Single Load:** The DINO ViT-S/16 model is loaded once in the metric’s constructor  
* **Correct Output:** `compute(...)` returns a sensible result (mean cosine similarity, optionally std/quantiles)  
* **Tests & Docs:** All existing and new unit tests pass, and the documentation is updated  

---

## ❓ Questions?
Feel free to jump into our Discord channel if you hit any roadblocks. Can’t wait to see your contribution of this metric! 🚀

### PR DESCRIPTION
**Pull request by Prashankavi**
## Related Issue

Fixes #(issue number)

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Additional Notes

Closes #318

- Added DinoScore metric using DINO ViT-S/16 embeddings
- Added unit tests
- Updated evaluation documentation

### CODE REVIEW

File: docs/user_manual/evaluate.rst [hunk 1]
```diff
@@ -501,6 +501,24 @@ Let's see how this works in code.
...
+            
+**dino_score**
+-----------------
+
+- **Description**: Measures the similarity between generated and reference images using DINO ViT-S/16 embeddings.
+- **Type**: Stateful metric
+- **Values**: Mean cosine similarity (float)
+- **Usage**:
+  
+  .. code-block:: python
```
Comment (johannaSommer) [inline docs/user_manual/evaluate.rst:513]: Great idea to add a code snippet here! Do you think it would make sense to change this to a code snippet for how the user can use this metric with the Evaluation Agent directly?

File: src/pruna/evaluation/metrics/metric_dino_score.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        emb_gt = F.normalize(emb_gt, dim=1)
+
+        # Compute cosine similarity
+        sim = (emb_x * emb_gt).sum(dim=1)
+        self.similarities.append(sim)
+
+    def compute(self):
+        sims = torch.cat(self.similarities)
+        mean_sim = sims.mean().item()
+        return {"name": self.metric_name, "value": mean_sim}
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_dino_score.py:45]: can we return a MetricResult here?

File: src/pruna/evaluation/metrics/metric_dino_score.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    metric_name = "dino_score"
+
+    def __init__(self, device="cuda"):
+        super().__init__()
+        self.device = device
+        # Load the DINO ViT-S/16 model once
+        self.model = timm.create_model("dino_vits16", pretrained=True)
+        self.model.eval().to(device)
+        # Add internal state to accumulate similarities
+        self.add_state("similarities", default=[], dist_reduce_fx="cat")
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_dino_score.py:18]: I see you used the torchmetrics notation of dist_reduce_fx there, but then never use it. I think we can remove it :)

File: src/pruna/evaluation/metrics/metric_dino_score.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import torch
+import torch.nn.functional as F
+import timm
+from pruna.evaluation.metrics.metric_stateful import StatefulMetric
+from pruna.evaluation.metrics.registry import MetricRegistry
+
+@MetricRegistry.register("dino_score")
+class DinoScore(StatefulMetric):
+    metric_name = "dino_score"
+
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_dino_score.py:10]: we should have a default call_type for the metric! From what I have seen you can use x_gt :)

File: src/pruna/evaluation/metrics/metric_dino_score.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import torch
+import torch.nn.functional as F
+import timm
+from pruna.evaluation.metrics.metric_stateful import StatefulMetric
+from pruna.evaluation.metrics.registry import MetricRegistry
+
+@MetricRegistry.register("dino_score")
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_dino_score.py:7]: I love that you used MetricRegistry here! Do you think it makes sense to dedicate a variable to the name and use that variable so we don't have to hardcode the name everywhere we use the name of this metric?

File: src/pruna/evaluation/metrics/metric_dino_score.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+class DinoScore(StatefulMetric):
+    metric_name = "dino_score"
+
+    def __init__(self, device="cuda"):
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_dino_score.py:11]: I'd rather we not hardcode the device here. pruna.engine.utils have `set_to_best_available` device which we can call with also None :)

File: src/pruna/evaluation/metrics/metric_dino_score.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        self.add_state("similarities", default=[], dist_reduce_fx="cat")
+
+    @torch.no_grad()
+    def update(self, x, gt, outputs=None):
+        """
+        x: list of generated images (tensors)
+        gt: list of reference images (tensors)
+        outputs: optional, ignored
+        """
+        x = torch.stack(x).to(self.device)
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_dino_score.py:27]: If we can use the metric_data_processor from the utils with then it would also handle device casting for you!
